assets/src/stories-editor/components/block-mover/block-draggable.js

```diff
@@ -54,13 +90,108 @@ BlockDraggable.propTypes = {
 	children: PropTypes.any.isRequired,
 	onDragStart: PropTypes.func,
 	onDragEnd: PropTypes.func,
+	onNeighborHover: PropTypes.func,
+	onNeighborDrop: PropTypes.func,
+	getNeighborPageId: PropTypes.func,
 };
 
-export default withSelect( ( select, { clientId } ) => {
-	const { getBlockIndex, getBlockRootClientId } = select( 'core/block-editor' );
-	const rootClientId = getBlockRootClientId( clientId );
-	return {
-		index: getBlockIndex( clientId, rootClientId ),
-		rootClientId,
-	};
-} )( BlockDraggable );
+export default compose(
+	withSelect( ( select, { clientId } ) => {
+		const { getBlockIndex, getBlockRootClientId, getBlockOrder, getBlock } = select( 'core/block-editor' );
+		const rootClientId = getBlockRootClientId( clientId );
+		return {
+			index: getBlockIndex( clientId, rootClientId ),
+			rootClientId,
+			block: getBlock( clientId ),
+			getBlockOrder,
+		};
+	} ),
+	withDispatch( ( dispatch, { getBlockOrder, rootClientId, clientId, block, blockName } ) => {
+		const { setCurrentPage } = dispatch( 'amp/story' );
+		const { selectBlock, removeBlock, insertBlock, updateBlockAttributes } = dispatch( 'core/block-editor' );
+
+		const isCTABlock = 'amp/amp-story-cta' === blockName;
+
+		/**
+		 * Is the current element allowed on a given page?
+		 *
+		 * This is currently only used to filter out instances, where CTA's aren't allowed.
+		 *
+		 * @param {number} targetPageIndex  Zero-based index of new page
+		 * @param {string} targetPageId  Block id of new page
+		 * @return {boolean} Returns true if this element is allowed on the given page, false otherwise.
+		 */
+		const isElementAllowedOnPage = ( targetPageIndex, targetPageId ) => {
```

Comment:
Might be worth comparing this against [ensureAllowedBlocksOnPaste](https://github.com/ampproject/amp-wp/blob/06ff84fec4357359ce556c889aacecfb06db66cb/assets/src/stories-editor/helpers/index.js#L1867-L1898).

Reply:
It does overlap, but it also handles blocks, that aren't draggable.

(also, it's a weird piece of code - why isn't it an `array.filter` expression?)

Reply:
Why not move this to helper functions file and then add some unit / integration tests?

( I didn't write that function, but I find switch statement easy to read. )

Reply:
The function internally use the `select`'ed function `getCallToActionBlock`, but that could of course be supplied to the helper. I'll look into it. But yes, it is a good candidate for unit test.

Reply:
Actually there's a lot of this logic existing elsewhere – e.g. used in the inserter, that only shows allowed elements and already hides CTA on first page or on pages with an existing CTA. I will try to combined all of these restrictions.

Reply:
This rewrite even caused me to remove a test, but I think it's okay as the heavy-lifting has been offloaded to Gutenberg core functionality in stead.

Also it's much cleaner now, thanks for the heads up!